File: week1/class2/tarea.py
from math import factorial
from typing import Tuple


def compute_tol(m) -> float:
    """Compute the tolerance for a given m

    Args:
        m (float): term to compute the tolerance

    Returns:
        float: computed tolerance
    """
    # Compute the current tolerance using the specified formula
    return 2**(2 * (m + 1)) / (factorial(m) * (2 * (m + 1) + 1))

def integral_example(tol: float) -> Tuple[float, int]:
    """Approximates an integral using a specified tolerance

    Args:
        tol (float): tolerance of the approximation

    Returns:
        Tuple[float, int]: Tuple with the approximation and the number of terms
    """
    # Initialize the approximation value
    aprox: float = 0.0
    # Initialize number of terms
    m: int = -1
    # Set the initial tolerance to a value greater than tol
    current_tol: float = 1.0

    # Loop to approximate the integral until the desired tolerance is achieved
    while current_tol > tol:
        # Increment m for the next term
        m += 1
        # Update the approximation using the series formula
        aprox += (-1)**m * 2**(2 * m) / (factorial(m) * (2 * m + 1))
        # Compute the current tolerance for the current m
        current_tol = compute_tol(m)

    # Return the approximation and the value of m
    return aprox, m


if __name__ == "__main__":
    # Define the tolerance
    tol: float = 0.000001
    # Call the 'integral_example' function and print the results
    print(integral_example(tol))

File: week1/class2/tarea.m
% Close all figures and clear the workspace
close all;
clear;

% Function to compute the current tolerance for a given m
function currentTol = computeTol(m)
  % Compute the current tolerance using the given formula
  currentTol = 2^(2 * (m + 1)) / (factorial(m) * (2 * (m + 1) + 1));
endfunction

% Function to approximate an integral using a specified tolerance
function [aprox, m] = integralExample(tol)
  % Initialize the approximation and m
  aprox = 0;
  m = -1;
  % Set initial tolerance to a value greater than tol
  currentTol = 1;

  % Loop to approximate the integral until the desired tolerance is achieved
  while (currentTol > tol)
    % Increment m for the next term
    m = m + 1;

    % Update the approximation using the series formula
    aprox = aprox + (-1)^m * 2^(2 * m) / (factorial(m) * (2 * m + 1));

    % Compute the current tolerance for the current m
    currentTol = computeTol(m);
  endwhile
endfunction

% Main script
% Call the 'integralExample' function with a tolerance of 0.000001
tol = 0.000001
[aprox, m] = integralExample(tol);

% Print the results
printf('Aprox = %f, m = %i', aprox, m);
